node/lib/util/checkout.js

```diff
@@ -81,10 +112,51 @@ exports.checkout = co.wrap(function *(metaRepo, committish) {
         }
 
         const repo = yield SubmoduleUtil.getRepo(metaRepo, name);
+        subRepos.push(repo);
         const sha = shas[name];
         yield GitUtil.fetchSha(repo, sha);
         const commit = yield repo.getCommit(sha);
-        yield GitUtil.setHeadHard(repo, commit);
+        subCommits.push(commit);
+        const error = yield dryRun(repo, commit);
+        if (null !== error) {
+            errors.push(
+             `Unable to checkout submodule ${colors.yellow(name)}: ${error}.`);
+        }
+    }));
+
+    // Throw an error if any dry-runs failed.
+
+    if (0 !== errors.length) {
+        throw new UserError(errors.join("\n"));
+    }
+
+    /**
+     * Checkout and set as head the specified `commit` in the specified `repo`.
+     */
+    const doCheckout = co.wrap(function *(repo, commit) {
+        yield NodeGit.Checkout.tree(repo, commit, {
+            checkoutStrategy: NodeGit.Checkout.STRATEGY.SAFE,
+        });
+        repo.setHeadDetached(commit);
+        yield NodeGit.Checkout.head(repo, {
+            checkoutStrategy: NodeGit.Checkout.STRATEGY.SAFE,
+        });
+    });
+
+    // Now do the actual checkouts.
+
+    yield doCheckout(metaRepo, commit);
```

Comment:
OK, I added some test cases to verify -- `checkout` does not view any *committed* changes in submodes as being unsafe.